File: news-server/app/api/v1/endpoints/stock.py
"""Stock News API Endpoints"""
from fastapi import APIRouter, Depends, Query
from sqlalchemy import text
from sqlalchemy.ext.asyncio import AsyncSession

from app.db.session import get_db
from app.core.logging import get_logger


logger = get_logger(__name__)
router = APIRouter()


@router.get("/latest")
async def get_latest_stock_news(
    limit: int = Query(default=10, le=50),
    db: AsyncSession = Depends(get_db),
):
    """Get latest stock news from database"""
    try:
        result = await db.execute(
            text("""
                SELECT 
                    content_hash as id,
                    title,
                    summary,
                    source_name,
                    category,
                    tickers,
                    sentiment,
                    impact_level,
                    processed_at as published_at
                FROM stock_news
                WHERE is_processed = TRUE
                ORDER BY processed_at DESC
                LIMIT :limit
            """),
            {"limit": limit}
        )
        rows = result.fetchall()
        
        items = []
        for row in rows:
            items.append({
                "id": row.id,
                "content_hash": row.id,
                "title": row.title,
                "summary": row.summary,
                "source_name": row.source_name,
                "category": row.category,
                "tickers": row.tickers or "",
                "sentiment": row.sentiment,
                "impact_level": row.impact_level,
                "published_at": row.published_at.isoformat() if row.published_at else None,
                "processed_at": row.published_at.isoformat() if row.published_at else None,
            })
        
        return {
            "items": items,
            "total": len(items),
        }
    except Exception as e:
        logger.error("Failed to fetch stock news", error=str(e))
        return {"items": [], "total": 0, "error": str(e)}

File: news-server/internal/api/stock.go
package api

import (
	"log/slog"
	"net/http"
	"strconv"

	"github.com/jackc/pgx/v5/pgxpool"
)

func RegisterStockRoutes(mux *http.ServeMux, db *pgxpool.Pool) {
	mux.HandleFunc("GET /api/v1/stock/latest", handleLatestStockNews(db))
}

func handleLatestStockNews(db *pgxpool.Pool) http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		limit, _ := strconv.Atoi(r.URL.Query().Get("limit"))
		if limit < 1 || limit > 50 {
			limit = 10
		}

		rows, err := db.Query(r.Context(),
			`SELECT content_hash, title, summary, source_name,
				category, tickers, sentiment, impact_level, processed_at
			 FROM stock_news
			 WHERE is_processed = TRUE
			 ORDER BY processed_at DESC
			 LIMIT $1`, limit)
		if err != nil {
			slog.Error("stock news query failed", "error", err)
			writeJSON(w, http.StatusInternalServerError, map[string]interface{}{"error": "query failed"})
			return
		}
		defer rows.Close()

		var items []map[string]interface{}
		for rows.Next() {
			var contentHash, title string
			var summary, sourceName, category, tickers, sentiment, impactLevel interface{}
			var processedAt interface{}

			if err := rows.Scan(&contentHash, &title, &summary, &sourceName,
				&category, &tickers, &sentiment, &impactLevel, &processedAt); err != nil {
				continue
			}

			items = append(items, map[string]interface{}{
				"id":           contentHash,
				"content_hash": contentHash,
				"title":        title,
				"summary":      summary,
				"source_name":  sourceName,
				"category":     category,
				"tickers":      tickers,
				"sentiment":    sentiment,
				"impact_level": impactLevel,
				"published_at": processedAt,
				"processed_at": processedAt,
			})
		}

		if items == nil {
			items = []map[string]interface{}{}
		}

		writeJSON(w, http.StatusOK, map[string]interface{}{
			"items": items,
			"total": len(items),
		})
	}
}
